crates/nu-std/std/testing.nu

```diff
@@ -0,0 +1,250 @@
+                '$context | after-each'
+            } else {
+                ''
+            }
+        }
+        | each {|test|
+            log info $"Running ($test.test) in module ($module.name) with context ($global_context)"
+            $test|insert result {|x|
+                run-test $test
+                | match $in.exit_code {
+                    0 => "pass",
+                    2 => "skip",
+                    _ => "fail",
+                }
+            }
+        }
+    )
+
+    if $module.after-all {
+        log info $"Running after-all for module ($module.name)"
+
+        run-test {
+                file: $module.file,
+                before-each: $"let context = ($global_context)",
+                after-each: '',
+                test: 'after-all'
+        }
+    }
+    return $tests
+}
+
+export def run-tests [
+    --path: path, # Path to look for tests. Default: current directory.
+    --module: string, # Test module to run. Default: all test modules found.
+    --test: string, # Individual test to run. Default: all test command found in the files.
+    --list, # list the selected tests without running them.
+] {
+    let module_search_pattern = ('**' | path join ({
+        stem: ($module | default "*")
+        extension: nu
+    } | path join))
+
+    let path = ($path | default $env.PWD)
+
+    if not ($path | path exists) {
+        throw-error {
+            msg: "directory_not_found"
+            label: "no such directory"
+            span: (metadata $path | get span)
+        }
+    }
+
+    if not ($module | is-empty) {
+        try { ls ($path | path join $module_search_pattern) | null } catch {
+            throw-error {
+                msg: "module_not_found"
+                label: $"no such module in ($path)"
+                span: (metadata $module | get span)
+            }
+        }
+    }
+
+    let modules = (
+        ls ($path | path join $module_search_pattern)
+        | each {|row| {file: $row.name name: ($row.name | path parse | get stem)}}
+        | upsert commands {|module|
+            get-commands $module.file
+        }
+        | upsert tests {|module| $module.commands|where $it starts-with "test_"}
+        | filter {|x| ($x.tests|length) > 0}
+        | filter {|x| if ($test|is-empty) {true} else {$test in $x.tests}}
+        | filter {|x| if ($module|is-empty) {true} else {$module == $x.name}}
+        | upsert before-each {|module| "before-each" in $module.commands}
+        | upsert before-all {|module| "before-all" in $module.commands}
+        | upsert after-each {|module| "after-each" in $module.commands}
+        | upsert after-all {|module| "after-all" in $module.commands}
+        | reject commands
+        | rename file name tests
+        | update tests {|x|
+            if ($test|is-empty) {
+                $x.tests
+            } else {
+                $x.tests
+                | where $it == $test
+            }
+        }
+    )
```

Comment:
here we might directly compute the `tests` column without the intermediate `commands` one :yum: 
```bash
    let modules = (
        ls ($path | path join $module_search_pattern)
        | each {|row| {
            file: $row.name
            name: ($row.name | path parse | get stem)
        }}
        | insert tests {|module|
            get-commands $module.file | where $it starts-with "test_"
        }
        | filter {|x| ($x.tests | length) > 0}
        | filter {|x| if ($test | is-empty) { true } else { $test in $x.tests }}
        | filter {|x| if ($module | is-empty) { true } else { $module == $x.name }}
        | upsert before-each {|module| "before-each" in $module.tests}
        | upsert before-all {|module| "before-all" in $module.tests}
        | upsert after-each {|module| "after-each" in $module.tests}
        | upsert after-all {|module| "after-all" in $module.tests}
        | rename file name tests
        | update tests {|x|
            if ($test | is-empty) {
                $x.tests
            } else {
                $x.tests | where $it == $test
            }
        }
    )
```
> **Note**
> there is a tiny bit of formatting too :relieved:

Reply:
This won't pick up before-each and after-each commands as they're not prefixed with test_ and that's what you're inserting

Reply:
this is a very good and valid point, thanks for the clarification about the `before_` and `after_` commands :relieved: